src/librustc_typeck/check/dropck.rs

```diff
@@ -450,45 +450,57 @@ fn iterate_over_potentially_unsafe_regions_in_type<'a, 'tcx>(
 
             let dtor_typescheme = ty::lookup_item_type(rcx.tcx(), impl_did);
             let dtor_generics = dtor_typescheme.generics;
-            let dtor_predicates = ty::lookup_predicates(rcx.tcx(), impl_did);
-
-            let has_pred_of_interest = dtor_predicates.predicates.iter().any(|pred| {
-                // In `impl<T> Drop where ...`, we automatically
-                // assume some predicate will be meaningful and thus
-                // represents a type through which we could reach
-                // borrowed data. However, there can be implicit
-                // predicates (namely for Sized), and so we still need
-                // to walk through and filter out those cases.
-
-                let result = match *pred {
-                    ty::Predicate::Trait(ty::Binder(ref t_pred)) => {
-                        let def_id = t_pred.trait_ref.def_id;
-                        match rcx.tcx().lang_items.to_builtin_kind(def_id) {
-                            Some(ty::BoundSend) |
-                            Some(ty::BoundSized) |
-                            Some(ty::BoundCopy) |
-                            Some(ty::BoundSync) => false,
-                            _ => true,
+
+            let mut has_pred_of_interest = false;
+
+            let mut seen_items = Vec::new();
```

Comment:
yeah I had a HashSet here originally when it was traversing every predicate.  But when I switched it to being focused on just the set of traits, I became somewhat confident that `n` would not get too large -- since I assume you would be unlikely to have a long chain of parent traits where *every* trait defined *no* items.